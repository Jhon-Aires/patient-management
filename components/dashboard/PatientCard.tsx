'use client';

import { useMemo } from 'react';
import { ChevronDown, ChevronUp, Edit, X } from 'lucide-react';

import { Patient } from '@/types/Patients';
import { Button } from '@/components/ui/button';
import {
  Card,
  CardContent,
  CardFooter,
  CardHeader,
  CardTitle
} from '@/components/ui/card';
import { cn } from '@/lib/utils';

const backupPatientAvatar = '/placeholder-user.jpg';

export default function PatientCard({
  patient,
  expanded,
  onExpand,
  onEdit,
  onDelete
}: {
  patient: Patient;
  expanded: boolean;
  onExpand: (id: number) => void;
  onEdit: (patient: Patient) => void;
  onDelete: (id: number) => void;
}) {
  const avatar = useMemo(() => {
    if (!patient.avatar) {
      return backupPatientAvatar;
    }

    // cloudflare-ipfs.com doesn't work, so I'm using ipfs.io instead
    if (patient.avatar.includes('cloudflare-ipfs.com')) {
      return patient.avatar.replace('cloudflare-ipfs.com', 'ipfs.io');
    }

    return patient.avatar;
  }, [patient.avatar]);

  return (
    <Card className="max-h-fit w-full">
      <CardHeader>
        <CardTitle className="flex justify-between items-center">
          <div className="flex items-center gap-2">
            <img
              src={avatar}
              alt={patient.name}
              className="h-10 w-10 rounded-full object-cover"
            />
            {patient.name}
          </div>
          <Button
            variant="ghost"
            size="sm"
            onClick={() => onExpand(patient.id)}
          >
            {expanded ? <ChevronUp /> : <ChevronDown />}
          </Button>
        </CardTitle>
        {/* <CardDescription>Age: {patient.age}</CardDescription> */}
      </CardHeader>
      <CardContent
        className={cn(`mb-4 overflow-hidden transition-all duration-300`, {
          'max-h-40': expanded,
          'max-h-0': !expanded,
          'overflow-y-scroll': expanded
        })}
      >
        <p>{patient.description}</p>
      </CardContent>
      <CardFooter className="flex justify-end space-x-2">
        <Button variant="outline" size="sm" onClick={() => onEdit(patient)}>
          <Edit className="mr-2 h-4 w-4" /> Edit
        </Button>
        <Button
          variant="destructive"
          size="sm"
          onClick={() => onDelete(patient.id)}
        >
          <X className="mr-2 h-4 w-4" /> Delete
        </Button>
      </CardFooter>
    </Card>
  );
}
