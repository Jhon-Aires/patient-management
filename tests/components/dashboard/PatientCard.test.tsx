import '@testing-library/jest-dom';

import { act, render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import PatientCard from '@/components/dashboard/PatientCard';

describe('PatientCard', () => {
  const patient = {
    avatar: '',
    createdAt: '',
    description: 'description',
    id: 1,
    name: 'Patient Name',
    website: ''
  };

  const onDelete = jest.fn();
  const onEdit = jest.fn();
  const onExpand = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const { asFragment } = render(
      <PatientCard
        expanded={false}
        onDelete={onDelete}
        onEdit={onEdit}
        onExpand={onExpand}
        patient={patient}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should call onExpand when the expand button is clicked', async () => {
    render(
      <PatientCard
        expanded={false}
        onDelete={onDelete}
        onEdit={onEdit}
        onExpand={onExpand}
        patient={patient}
      />
    );

    const heading = screen.getByRole('heading', {
      name: /patient name patient name/i
    });

    const expandButton = within(heading).getByRole('button');

    await userEvent.click(expandButton);

    expect(onExpand).toHaveBeenCalledWith(patient.id);
  });

  it('should call onEdit when the edit button is clicked', async () => {
    render(
      <PatientCard
        expanded={false}
        onDelete={onDelete}
        onEdit={onEdit}
        onExpand={onExpand}
        patient={patient}
      />
    );

    const editButton = screen.getByRole('button', {
      name: /edit/i
    });

    await userEvent.click(editButton);

    expect(onEdit).toHaveBeenCalledWith(patient);
  });

  it('should call onDelete when the delete button is clicked', async () => {
    render(
      <PatientCard
        expanded={false}
        onDelete={onDelete}
        onEdit={onEdit}
        onExpand={onExpand}
        patient={patient}
      />
    );

    const deleteButton = screen.getByRole('button', {
      name: /delete/i
    });

    await userEvent.click(deleteButton);

    expect(onDelete).toHaveBeenCalledWith(patient.id);
  });
});
