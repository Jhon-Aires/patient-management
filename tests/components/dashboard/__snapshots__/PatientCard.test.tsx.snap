// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PatientCard should render correctly 1`] = `
<DocumentFragment>
  <div
    class="rounded-lg border bg-card text-card-foreground shadow-sm max-h-fit w-full"
  >
    <div
      class="flex flex-col space-y-1.5 p-6"
    >
      <h3
        class="text-2xl font-semibold leading-none tracking-tight flex justify-between items-center"
      >
        <div
          class="flex items-center gap-2"
        >
          <img
            alt="Patient Name"
            class="h-10 w-10 rounded-full object-cover"
            src="/placeholder-user.jpg"
          />
          Patient Name
        </div>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-9 rounded-md px-3"
        >
          <svg
            class="lucide lucide-chevron-down"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </h3>
    </div>
    <div
      class="p-6 pt-0 mb-4 overflow-hidden transition-all duration-300 max-h-0"
    >
      <p>
        description
      </p>
    </div>
    <div
      class="items-center p-6 pt-0 flex justify-end space-x-2"
    >
      <button
        class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-9 rounded-md px-3"
        data-testid="edit-button"
      >
        <svg
          class="lucide lucide-square-pen mr-2 h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
          />
          <path
            d="M18.375 2.625a1 1 0 0 1 3 3l-9.013 9.014a2 2 0 0 1-.853.505l-2.873.84a.5.5 0 0 1-.62-.62l.84-2.873a2 2 0 0 1 .506-.852z"
          />
        </svg>
         Edit
      </button>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-destructive text-destructive-foreground hover:bg-destructive/90 h-9 rounded-md px-3"
      >
        <svg
          class="lucide lucide-x mr-2 h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
         Delete
      </button>
    </div>
  </div>
</DocumentFragment>
`;
